kwave/__init__.py

```diff
@@ -1,22 +1,123 @@
 import os
-from os import environ
 import sys
+import urllib.request
+from os import environ
 
 from kwave.kgrid import kWaveGrid
 from kwave.kmedium import kWaveMedium
-from kwave.options import SimulationOptions
 from kwave.ktransducer import NotATransducer
+from kwave.options import SimulationOptions
 
 VERSION = '0.1.0'
 # Set environment variable to binaries to get rid of user warning
 # This code is a crutch and should be removed when kspaceFirstOrder
 # is refactored
 
-if sys.platform.startswith('darwin'):
-    system = "macOS"
-elif sys.platform.startswith('linux'):
-    system = "linux"
-elif "win" in sys.platform:
-    system = "windows"
+# Get current system type
+system = sys.platform
+
+binary_path = os.path.join(os.getcwd(), 'kwave', 'bin', system)
+environ['KWAVE_BINARY_PATH'] = binary_path
+
+url_base = "https://github.com/waltsims/"
+
+prefix = "https://github.com/waltsims/kspaceFirstOrder-{0}-{1}/releases/download/v1.3.0/"
+
+common_filenames = ["cufft64_10.dll", "hdf5.dll", "hdf5_hl.dll", "libiomp5md.dll", "libmmd.dll",
+                    "msvcp140.dll", "svml_dispmd.dll", "szip.dll", "vcruntime140.dll", "zlib.dll"]
+
+specific_omp_filenames = ["kspaceFirstOrder-OMP.exe"]
+specific_cuda_filenames = ["kspaceFirstOrder-CUDA.exe"]
+
+
+def binaries_present() -> bool:
+    """
+    Check if binaries are present
+    Returns:
+        bool, True if binaries are present, False otherwise
+
+    """
+    binary_list = {
+        "linux": [
+            # "acousticFieldPropagator-OMP",
+            "kspaceFirstOrder-OMP",
+            "kspaceFirstOrder-CUDA"
+        ],
+        "darwin": [
+            # "acousticFieldPropagator-OMP",
+            "kspaceFirstOrder-OMP",
+            "kspaceFirstOrder-CUDA"
+        ],
+        "cygwin": specific_omp_filenames + specific_cuda_filenames + common_filenames
+    }
+
+    for binary in binary_list[system]:
+        if not os.path.exists(os.path.join(binary_path, binary)):
+            print(f"{binary} not found")
+            return False
+    return True
+
+
+def download_binaries(system_os: str, bin_type: str):
+    """
+    Download binary from release url
+    Args:
+        system_os: string, current system type
+        bin_type: string of "OMP" or "CUDA"
+
+    Returns:
+        None
+
+    """
+
+    # GitHub release URLs
+    url_dict = {
+        "linux": {
+            "cuda": [url_base + "kspaceFirstOrder-CUDA-linux/releases/download/v1.3/kspaceFirstOrder-CUDA"],
+            "cpu": [url_base + "kspaceFirstOrder-OMP-linux/releases/download/v1.3.0/kspaceFirstOrder-OMP"],
+        },
+        # "darwin": {
```

Comment:
Consider removing the comments if we don't support lazy-installation in `darwin` yet